File: NeetCode 150/Graphs/1584. Min Cost to Connect All Points.py
class Solution:
    def minCostConnectPoints(self, points: List[List[int]]) -> int:
        #Solution : BFS + MinHeap
        n = len(points)
        #build adj list
        adj = {i:[] for i in range(n)}
        for i in range(n):
            x1,y1 = points[i]
            for j in range(i+1,n):
                x2,y2 = points[j]
                dist = abs(x1-x2) + abs(y1-y2)

                adj[i].append((dist,j))
                adj[j].append((dist,i))
        #Prim's Algo
        res = 0
        visit = set()
        min_heap = [(0,0)] #cost,point
        while len(visit) < n: #ensure the condition is diff and we are not checking min_heap instead
            cost,i = heapq.heappop(min_heap)
            if i in visit:
                continue
            res += cost #we are adding the cost to the result only here
            visit.add(i)
            for neiCost,nei in adj[i]:
                #If already in visit. added to res already, so don't add
                if nei not in visit:
                    heapq.heappush(min_heap,(neiCost,nei)) 
                    #we are only adding curr dist and not sum with previous.
                    
        return res

File: NeedCode 150 (JAVA)/Graphs/1584. Min Cost to Connect All Points.java
/*Prim’s algorithm grows the MST one node at a time.

At each step:

pick the not-yet-used point with the smallest connection cost
add it to the MST
update the best known cost for the remaining points
*/
class Solution {
    public int minCostConnectPoints(int[][] points) {
        int n = points.length;

        PriorityQueue<int[]> pq = new PriorityQueue<>((a,b) -> a[0] - b[0]);
        boolean[] visited = new boolean[n];

        pq.offer(new int[]{0,0}); //pq.offer([0,0]) will not work. it is python
        //visited[0] = true; Note, do not add visited here. Visited should only be update in while loop and after polling from PQ. Or else the Points Connection will never start

        int total_cost = 0;
        int count = 0; //since point 0 is not included, we are not adding +1 here

        while(!pq.isEmpty() && count < n){
            int[] curr = pq.poll();
            int cost = curr[0];
            int point = curr[1];

            if(visited[point])
                continue;
            
            visited[point] = true;
            total_cost += cost;
            count++;

            //Since the graph is complete (every point can connect to every other point), you don't build an adjacency list. When you remove a point from the heap, you compute its distance to every unvisited point on the fly. Loop is from 0 to n
            for(int next = 0; next<n;next++){
                if(!visited[next]){
                    int dist = Math.abs(points[point][0] - points[next][0]) + Math.abs(points[point][1] - points[next][1]);

                    pq.offer(new int[]{dist,next});
                }
            }
        }

        return total_cost; 
    }
}
